### DISCUSSION TITLE
Perfect TornadoVM Environment.

### DISCUSSION THREAD

**Discussion by bradmesserle** [beehive-lab/TornadoVM#73] [General]
It sounds like the apple setup is not compatible.
So.. if you were to going start fresh. .what would be the perfect hardware / software environment?

Thank you!

**Comment by jjfumero** [comment]
TornadoVM can run on OSx, although we currently have some issues: https://github.com/beehive-lab/TornadoVM/issues/66 

Apart from that, it should work well on OSx. 

I personally do not have Mac but if you have, and you want to contribute to the project, we are happy to guide you to fix those issues. 

**Regarding hardware**:
* GPUs: We develop TornadoVM in  "standard" laptops (NVIDIA GPUs + Intel HD Graphics). We can also run on AMD GPUs. However, I will need to check on the latest ROCm drivers.  Besides,  it has been pointed to us some errors regarding events, but we have a workaround for this while we fix it. We are in the process of redesigning the event-handler within TornadoVM. 

BTW, this article by Gary Frost shows the potential of TornadoVM on AMD GPUs: https://e2data.eu/blog/java-gpu-accelerated-viola-jones-face-detection-with-tornadovm 

* FPGAs: If you are interested in FPGAs, we currently develop on Intel Nallatech 385a and Xilinx KCU1500 with SDAccel  (These FPGAs are connected to standards Intel i7 PCs in our lab). 


**Regarding software:**
* The TornadoVM team develop on Linux (CentOS >= 7.X and Ubuntu >= 18.x ). 
* Recently, some members of the team are also using Windows. 
* IDE: We currently use IntelliJ, although I personally used Eclipse for a long time in TornadoVM. So any of those should work well. 

**Supported OpenCL & PTX Drivers:** 
- This doc shows an updated list of drivers that we have checked with TornadoVM: https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/TORNADOVM_TESTED_DRIVERS.md 

**Supported platforms:** 
- https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/1_INSTALL.md#installing-tornadovm

**Reply by bradmesserle** [reply]
Are there any SoC boards that have been tested?

Like https://www.amazon.com/dp/B0797P9F1Z?tag=amz-mkt-chr-us-20&ascsubtag=1ba00-01000-org00-mac00-other-nomod-us000-pcomp-feature-scomp-wm-5&ref=aa_scomp

My use case might be more suited for fpga.. then gpu's..

thank you!

**Reply by stratika** [reply]
We have tested it on the HiKey960 board with Arm Mali GPUs: https://www.96boards.org/product/hikey960/. 

We have not tested TornadoVM on the Arm-Xilinx FPGA SoC, yet. However, I believe it should not be difficult to run because:
1) TornadoVM can run on Arm64 via GraalVM JDK 11. (I think Zynq-7000 has the Arm Cortex A9 which is Armv7 ISA. This will not work.)
2) We have recently integrated support for the Xilinx Vitis HLS.

But this is with any reservation, as we have not tested it yet.

**Reply by bradmesserle** [reply]
What about this chip?

https://www.xilinx.com/products/silicon-devices/soc/zynq-ultrascale-mpsoc.html

its armA53.. 
https://www.xilinx.com/products/boards-and-kits/1-vad4rl.html
Looking at purchasing this guy..

thx

**Reply by stratika** [reply]
1. Regarding the Arm cores, Arm Cortex A53 should be fine for TornadoVM to run.

2. Regarding the selection of a board, I would double check the platforms supported by Vitis. Ultra96-V2 may be also supported, but I have no experience with this board. The Vitis supported platforms are listed here: https://www.xilinx.com/html_docs/xilinx2020_2/vitis_doc/embedded_release_notes.html#nwp1602134447203

Also, take into consideration that the FPGA in the Avnet Ultra96-V2 board seems to be smaller compared to the one in the ZCU102 development board. This decision should be aligned with your applications' requirements. 

This is the selection guides from Xilinx:
https://www.xilinx.com/support/documentation/selection-guides/zynq-ultrascale-plus-product-selection-guide.pdf#EG

Ultra96-V2 uses the ZU3EG, while ZCU102 uses the ZU9EG device.

Hope this helps. :-)

**Reply by jjfumero** [reply]
Also, keep in mind that to run TornadoVM on FPGAs, you need OpenCL support for the selected board/s.

**Reply by bradmesserle** [reply]
Yea i need something inexpensive to play around with..  If I get everything working on the cheepe chinese board I will share the details.. and i am sure i will be posting questions trying to get it to work :)


Thanks!

**Reply by bradmesserle** [reply]
Looks like there other other boards as well.

https://www.xilinx.com/support/documentation/selection-guides/zynq-mpsoc-kit-selection-guide.pdf
